src/algebra/group/inj_surj.lean

```diff
@@ -44,6 +44,16 @@ protected def semigroup [semigroup M₂] (f : M₁ → M₂) (hf : injective f)
 { mul_assoc := λ x y z, hf $ by erw [mul, mul, mul, mul, mul_assoc],
   ..‹has_mul M₁› }
 
+/-- A type endowed with `0` and `*` is a `semigroup_with_zero`,
+if it admits an injective map that preserves `0` and `*` to a `semigroup_with_zero`. -/
+protected def semigroup_with_zero
+  [has_zero M₁] [semigroup_with_zero M₂] (f : M₁ → M₂) (hf : injective f)
+  (zero : f 0 = 0) (mul : ∀ x y, f (x * y) = f x * f y) :
+  semigroup_with_zero M₁ :=
+{ zero_mul := λ x, hf $ by erw [mul, zero, zero_mul],
+  mul_zero := λ x, hf $ by erw [mul, zero, mul_zero],
+  ..‹has_zero M₁›, .. hf.semigroup f mul }
+
```

Comment:
I think this belongs in a different file next to the one for `monoid_with_zero`:

https://github.com/leanprover-community/mathlib/blob/f29b0b49badab1bcbe338cb79d102e36e79dce09/src/algebra/group_with_zero/basic.lean#L183-L213

Reply:
What is the logic here?

I see that `algebra/group_with_zero/basic.lean` contains these inj/surj lemms for:

- `mul_zero_class`
- `no_zero_divisors`
- `monoid_with_zero`
- `comm_monoid_with_zero`
- `cancel_monoid_with_zero`
- `group_with_zero`
- `group_with_zero_div`
- `comm_group_with_zero`
- `comm_group_with_zero_div`

whereas `algebra/group/inj_surj.lean` contains such lemmas for:

- `semigroup`
- `semigroup_with_zero` <-- the lemma in question here
- `comm_semigroup`
- `left_cancel_semigroup`
- `right_cancel_semigroup`
- `mul_one_class`
- `monoid`
- `left_cancel_monoid`
- `comm_monoid`
- `div_inv_monoid`
- `group`
- `group_div`
- `comm_group`
- `comm_group_div`

Reply:
Never mind, I think I see the motivation now.